compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -734,6 +736,58 @@ pub struct CapturedPlace<'tcx> {
     pub info: CaptureInfo<'tcx>,
 }
 
+pub fn place_to_string_for_capture(tcx: TyCtxt<'tcx>, place: &HirPlace<'tcx>) -> String {
+    let name = match place.base {
+        HirPlaceBase::Upvar(upvar_id) => tcx.hir().name(upvar_id.var_path.hir_id).to_string(),
+        _ => bug!("Capture_information should only contain upvars"),
+    };
+    let mut curr_string = format!("{}", name);
+
+    for (i, proj) in place.projections.iter().enumerate() {
+        if HirProjectionKind::Deref == proj.kind {
+            curr_string = format!("*{}", curr_string);
+        } else {
+            match place.ty_before_projection(i).kind() {
+                ty::Adt(def, ..) if def.is_enum() => {
+                    // we might be projecting *to* a variant, or to a field *in* a variant.
```

Comment:
Actually, I'd combine the is_enum() and the non enum arm. And assert that `def.variants.len() == 1`. Leave a comment saying no projection should be appplied to multi-variant enums and they should be completely captured.

Reply:
I don't think it makes sense to do that honeslty.  The way we access the variant is different for enum and non enum so we will need to have an if else condition regardless

Reply:
the non_enum_variant does the same thing, as `def.variants[variant]` because for non_enum_variant just uses `VariantIdx::new(0)`, which would be the `idx` in `Field(idx, variant)`.